workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs

```diff
@@ -126,117 +127,108 @@ public RedirectedProcess RedirectOutputOptions(OutputRedirectBehaviour redirectB
+                        processStandardOutput.Add(output);
                     }
-                };
-
-                process.ErrorDataReceived += (sender, args) =>
+                });
+                AddErrorProcessing(error =>
                 {
-                    var errorString = args.Data;
-                    if (string.IsNullOrEmpty(errorString))
+                    lock (processStandardOutput)
                     {
-                        return;
+                        processStandardError.Add(error);
                     }
+                });
 
-                    if ((outputRedirectBehaviour & OutputRedirectBehaviour.ProcessSpatialOutput) != OutputRedirectBehaviour.None)
-                    {
-                        errorString = ProcessSpatialOutput(errorString);
-                    }
+                if (token == null)
+                {
+                    var exitCode = Run();
 
-                    if ((outputRedirectBehaviour & OutputRedirectBehaviour.RedirectStdErr) !=
-                        OutputRedirectBehaviour.None)
+                    return new RedirectedProcessResult
                     {
-                        Debug.LogError(errorString);
-                    }
+                        ExitCode = exitCode,
+                        Stdout = processStandardOutput,
+                        Stderr = processStandardError
+                    };
+                }
 
-                    if (outputLog != null)
-                    {
-                        lock (outputLog)
-                        {
-                            outputLog.AppendLine(ProcessSpatialOutput(errorString));
-                        }
-                    }
+                var task = RunWithCancel(token.Value);
+                task.Wait();
 
-                    foreach (var errorProcessor in errorProcessors)
-                    {
-                        try
-                        {
-                            errorProcessor.Invoke(errorString);
-                        }
-                        catch (Exception e)
-                        {
-                            Debug.LogException(e);
-                        }
-                    }
+                return new RedirectedProcessResult
+                {
+                    ExitCode = task.Result,
+                    Stdout = processStandardOutput,
+                    Stderr = processStandardError
                 };
+            });
+        }
 
-                process.EnableRaisingEvents = true;
+        private async Task<int> RunWithCancel(CancellationToken token)
+        {
+            var (process, outputLog) = SetupProcess();
 
-                if (!process.Start())
+            using (process)
+            {
+                await Task.Run(() =>
                 {
-                    throw new Exception(
-                        $"Failed to run {info.FileName} {info.Arguments}");
-                }
+                    while (!process.HasExited)
```

Comment:
This polling approach is an anti-pattern.
Use the `Process.Exited` event and a `TaskCompletionSource` to signal completion, and the `CancellationToken`'s `Register` method to register a callback that kills the process on cancellation.

Reply:
PTAL and double check my work 😄